File: parallel-algorithms/graph/parallel_bfs.go
/*
Parallel Breadth-First Search (BFS) Implementation in Go

This implementation leverages Go's goroutines and channels for efficient parallel graph traversal.

Features:
- Level-synchronous parallel BFS
- Channel-based work distribution
- Concurrent visited tracking with sync.Map
- Multiple strategies (level-sync, worker pool)

Time Complexity: O(V + E) where V = vertices, E = edges
Space Complexity: O(V)

Author: Algorithms Multiverse
*/

package main

import (
	"fmt"
	"math/rand"
	"runtime"
	"sync"
	"time"
)

// Graph represents an adjacency list graph
type Graph struct {
	adjList    map[int][]int
	numVertices int
	numEdges   int
	directed   bool
}

// NewGraph creates a new graph
func NewGraph(directed bool) *Graph {
	return &Graph{
		adjList:  make(map[int][]int),
		directed: directed,
	}
}

// AddEdge adds an edge to the graph
func (g *Graph) AddEdge(u, v int) {
	g.adjList[u] = append(g.adjList[u], v)
	if !g.directed {
		g.adjList[v] = append(g.adjList[v], u)
	}

	if u >= g.numVertices {
		g.numVertices = u + 1
	}
	if v >= g.numVertices {
		g.numVertices = v + 1
	}
	g.numEdges++
}

// GetNeighbors returns neighbors of a vertex
func (g *Graph) GetNeighbors(vertex int) []int {
	return g.adjList[vertex]
}

// BFSResult contains results from BFS traversal
type BFSResult struct {
	VisitedOrder  []int
	Distances     map[int]int
	Parent        map[int]int
	TimeTaken     time.Duration
	VerticesProcessed int
	EdgesExplored int
	Method        string
	NumWorkers    int
}

// ParallelBFS handles parallel BFS operations
type ParallelBFS struct {
	graph       *Graph
	numWorkers  int
	visited     sync.Map
	distances   sync.Map
	parent      sync.Map
	visitedOrder []int
	orderMutex  sync.Mutex
	edgesExplored int64
	edgesMutex   sync.Mutex
}

// NewParallelBFS creates a new parallel BFS instance
func NewParallelBFS(graph *Graph, numWorkers int) *ParallelBFS {
	if numWorkers <= 0 {
		numWorkers = runtime.NumCPU()
	}

	return &ParallelBFS{
		graph:      graph,
		numWorkers: numWorkers,
	}
}

// reset clears all state
func (pbfs *ParallelBFS) reset() {
	pbfs.visited = sync.Map{}
	pbfs.distances = sync.Map{}
	pbfs.parent = sync.Map{}
	pbfs.visitedOrder = nil
	pbfs.edgesExplored = 0
}

// addToVisitedOrder adds a vertex to visited order (thread-safe)
func (pbfs *ParallelBFS) addToVisitedOrder(vertex int) {
	pbfs.orderMutex.Lock()
	pbfs.visitedOrder = append(pbfs.visitedOrder, vertex)
	pbfs.orderMutex.Unlock()
}

// incrementEdges increments edge counter (thread-safe)
func (pbfs *ParallelBFS) incrementEdges(count int64) {
	pbfs.edgesMutex.Lock()
	pbfs.edgesExplored += count
	pbfs.edgesMutex.Unlock()
}

// SequentialBFS performs standard sequential BFS
func (pbfs *ParallelBFS) SequentialBFS(start int) BFSResult {
	pbfs.reset()
	startTime := time.Now()

	queue := []int{start}
	pbfs.visited.Store(start, true)
	pbfs.distances.Store(start, 0)
	pbfs.parent.Store(start, -1)
	pbfs.addToVisitedOrder(start)

	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]

		distance, _ := pbfs.distances.Load(vertex)

		for _, neighbor := range pbfs.graph.GetNeighbors(vertex) {
			pbfs.incrementEdges(1)

			if _, visited := pbfs.visited.Load(neighbor); !visited {
				pbfs.visited.Store(neighbor, true)
				pbfs.distances.Store(neighbor, distance.(int)+1)
				pbfs.parent.Store(neighbor, vertex)
				pbfs.addToVisitedOrder(neighbor)
				queue = append(queue, neighbor)
			}
		}
	}

	return pbfs.buildResult(time.Since(startTime), "sequential", 1)
}

// ParallelBFSLevelSync performs level-synchronous parallel BFS
func (pbfs *ParallelBFS) ParallelBFSLevelSync(start int) BFSResult {
	pbfs.reset()
	startTime := time.Now()

	currentLevel := []int{start}
	pbfs.visited.Store(start, true)
	pbfs.distances.Store(start, 0)
	pbfs.parent.Store(start, -1)
	pbfs.addToVisitedOrder(start)

	level := 0

	for len(currentLevel) > 0 {
		// Channel to collect next level vertices
		nextLevelChan := make(chan int, len(currentLevel)*10)
		var wg sync.WaitGroup

		// Process current level in parallel
		for _, vertex := range currentLevel {
			wg.Add(1)
			go func(v int) {
				defer wg.Done()

				localEdges := int64(0)
				for _, neighbor := range pbfs.graph.GetNeighbors(v) {
					localEdges++

					// Try to mark as visited
					if _, visited := pbfs.visited.LoadOrStore(neighbor, true); !visited {
						pbfs.distances.Store(neighbor, level+1)
						pbfs.parent.Store(neighbor, v)
						nextLevelChan <- neighbor
					}
				}
				pbfs.incrementEdges(localEdges)
			}(vertex)
		}

		// Wait for all goroutines to finish
		wg.Wait()
		close(nextLevelChan)

		// Collect next level
		currentLevel = nil
		for vertex := range nextLevelChan {
			currentLevel = append(currentLevel, vertex)
			pbfs.addToVisitedOrder(vertex)
		}

		level++
	}

	return pbfs.buildResult(time.Since(startTime), "level_sync", pbfs.numWorkers)
}

// ParallelBFSWorkerPool uses worker pool pattern
func (pbfs *ParallelBFS) ParallelBFSWorkerPool(start int) BFSResult {
	pbfs.reset()
	startTime := time.Now()

	// Channels for work distribution
	workQueue := make(chan int, pbfs.numWorkers*2)
	done := make(chan bool)

	// Initialize
	pbfs.visited.Store(start, true)
	pbfs.distances.Store(start, 0)
	pbfs.parent.Store(start, -1)
	pbfs.addToVisitedOrder(start)
	workQueue <- start

	// Track active workers
	var activeWorkers int32
	var workerMutex sync.Mutex

	// Worker function
	worker := func() {
		for {
			select {
			case vertex, ok := <-workQueue:
				if !ok {
					return
				}

				workerMutex.Lock()
				activeWorkers++
				workerMutex.Unlock()

				distance, _ := pbfs.distances.Load(vertex)
				localEdges := int64(0)

				for _, neighbor := range pbfs.graph.GetNeighbors(vertex) {
					localEdges++

					if _, visited := pbfs.visited.LoadOrStore(neighbor, true); !visited {
						pbfs.distances.Store(neighbor, distance.(int)+1)
						pbfs.parent.Store(neighbor, vertex)
						pbfs.addToVisitedOrder(neighbor)

						select {
						case workQueue <- neighbor:
						case <-time.After(10 * time.Millisecond):
							// Queue might be full, skip
						}
					}
				}

				pbfs.incrementEdges(localEdges)

				workerMutex.Lock()
				activeWorkers--
				workerMutex.Unlock()

			case <-time.After(100 * time.Millisecond):
				// Check if all work is done
				workerMutex.Lock()
				if activeWorkers == 0 && len(workQueue) == 0 {
					workerMutex.Unlock()
					done <- true
					return
				}
				workerMutex.Unlock()
			}
		}
	}

	// Start workers
	for i := 0; i < pbfs.numWorkers; i++ {
		go worker()
	}

	// Wait for completion
	<-done
	close(workQueue)

	return pbfs.buildResult(time.Since(startTime), "worker_pool", pbfs.numWorkers)
}

// buildResult creates BFSResult from current state
func (pbfs *ParallelBFS) buildResult(duration time.Duration, method string, workers int) BFSResult {
	distances := make(map[int]int)
	parent := make(map[int]int)

	pbfs.distances.Range(func(key, value interface{}) bool {
		distances[key.(int)] = value.(int)
		return true
	})

	pbfs.parent.Range(func(key, value interface{}) bool {
		parent[key.(int)] = value.(int)
		return true
	})

	return BFSResult{
		VisitedOrder:      pbfs.visitedOrder,
		Distances:         distances,
		Parent:            parent,
		TimeTaken:         duration,
		VerticesProcessed: len(pbfs.visitedOrder),
		EdgesExplored:     int(pbfs.edgesExplored),
		Method:            method,
		NumWorkers:        workers,
	}
}

// createRandomGraph creates a random graph for testing
func createRandomGraph(numVertices, edgesPerVertex int) *Graph {
	graph := NewGraph(false)

	for v := 0; v < numVertices; v++ {
		for i := 0; i < edgesPerVertex; i++ {
			neighbor := rand.Intn(numVertices)
			if neighbor != v {
				graph.AddEdge(v, neighbor)
			}
		}
	}

	return graph
}

// benchmarkAllMethods benchmarks all BFS methods
func benchmarkAllMethods(graph *Graph, start int) map[string]BFSResult {
	results := make(map[string]BFSResult)

	methods := []string{"sequential", "level_sync", "worker_pool"}

	for _, method := range methods {
		bfs := NewParallelBFS(graph, runtime.NumCPU())

		var result BFSResult
		switch method {
		case "sequential":
			result = bfs.SequentialBFS(start)
		case "level_sync":
			result = bfs.ParallelBFSLevelSync(start)
		case "worker_pool":
			result = bfs.ParallelBFSWorkerPool(start)
		}

		results[method] = result
	}

	return results
}

// calculateSpeedup calculates speedup factor
func calculateSpeedup(seqTime, parallelTime time.Duration) float64 {
	if parallelTime == 0 {
		return 0
	}
	return float64(seqTime) / float64(parallelTime)
}

// repeatString repeats a string n times
func repeatString(s string, count int) string {
	result := ""
	for i := 0; i < count; i++ {
		result += s
	}
	return result
}

func main() {
	fmt.Println(repeatString("=", 80))
	fmt.Println("PARALLEL BREADTH-FIRST SEARCH (BFS) - Go Implementation")
	fmt.Println(repeatString("=", 80))
	fmt.Printf("Number of CPUs available: %d\n", runtime.NumCPU())

	// Test with different graph sizes
	testCases := []struct {
		numVertices    int
		edgesPerVertex int
	}{
		{1000, 5},
		{5000, 10},
		{10000, 15},
	}

	for _, tc := range testCases {
		fmt.Printf("\n%s\n", repeatString("=", 80))
		fmt.Printf("Testing with %d vertices, ~%d edges per vertex\n",
			tc.numVertices, tc.edgesPerVertex)
		fmt.Printf("%s\n", repeatString("=", 80))

		// Create random graph
		graph := createRandomGraph(tc.numVertices, tc.edgesPerVertex)
		fmt.Printf("Graph: %d vertices, ~%d edges\n",
			graph.numVertices, graph.numEdges)

		// Benchmark all methods
		results := benchmarkAllMethods(graph, 0)

		// Display results
		fmt.Printf("\n%-20s %-15s %-15s %-10s\n",
			"Method", "Time", "Vertices", "Speedup")
		fmt.Println(repeatString("-", 80))

		var seqTime time.Duration
		if result, ok := results["sequential"]; ok {
			seqTime = result.TimeTaken
		}

		for _, method := range []string{"sequential", "level_sync", "worker_pool"} {
			if result, ok := results[method]; ok {
				speedup := calculateSpeedup(seqTime, result.TimeTaken)

				fmt.Printf("%-20s %13s  %13d  %8.2fx\n",
					method,
					result.TimeTaken,
					result.VerticesProcessed,
					speedup)
			}
		}
	}

	fmt.Printf("\n%s\n", repeatString("=", 80))
	fmt.Println("WHEN TO USE PARALLEL BFS IN GO")
	fmt.Printf("%s\n", repeatString("=", 80))
	fmt.Println(`
Parallel BFS with goroutines is beneficial when:

✓ Large graphs (>10,000 vertices)
✓ High branching factor (many neighbors per vertex)
✓ Multiple CPU cores available
✓ Graph fits in memory

Go-specific advantages:
✓ Lightweight goroutines (2-8KB stack)
✓ Efficient channel communication
✓ sync.Map for concurrent visited tracking
✓ Built-in CSP concurrency model
✓ Easy to reason about with goroutines and channels

Key strategies:

1. Level-Synchronous (Recommended)
   - Maintains exact BFS ordering
   - Barrier synchronization between levels
   - Best for correctness

2. Worker Pool
   - Continuous work distribution
   - Better load balancing
   - May not maintain exact order

Performance characteristics:
- Speedup limited by graph diameter (sequential depth)
- Best for "wide" graphs (high branching factor)
- Worse for "deep" graphs (low branching factor)
- Goroutine overhead: ~200ns per spawn
- Channel overhead: ~100ns per operation

Challenges addressed:
- Thread contention: Using sync.Map for lock-free reads
- Load imbalance: Level synchronization or worker pools
- Memory bandwidth: Efficient data structures
- Goroutine explosion: Controlled spawning

Best practices:
1. Use level-synchronous for exact BFS ordering
2. Use worker pool for better load balance
3. Use sync.Map for concurrent visited tracking
4. Buffer channels appropriately (2x-10x workers)
5. Profile with pprof to identify bottlenecks

Applications:
- Social network analysis
- Web crawling
- Network routing
- Shortest path finding
- Connected components

Debugging and profiling:
- go test -race: Detect race conditions
- pprof: CPU and memory profiling
- go tool trace: Visualize goroutines
- GODEBUG=schedtrace=1000: Scheduler details
	`)

	fmt.Println("\nDemonstration complete!")
}

File: parallel-algorithms/graph/parallel_bfs.py
"""
Parallel Breadth-First Search (BFS) Implementation

This module implements parallel BFS with level synchronization for graph traversal.

Features:
- Level-synchronous parallel BFS
- Thread-safe visited set using locks
- Work distribution across multiple workers
- Support for both directed and undirected graphs
- Distance computation and path reconstruction

Time Complexity: O(V + E) where V = vertices, E = edges
Space Complexity: O(V)
Speedup: Depends on graph structure and branching factor

When to use parallel BFS:
- Large graphs (>10,000 vertices)
- High branching factor
- Multi-core systems available
- Level-wise processing needed

Author: Algorithms Multiverse
"""

import concurrent.futures
import multiprocessing as mp
import threading
import time
import queue
from typing import List, Set, Dict, Tuple, Optional
from dataclasses import dataclass
from collections import defaultdict, deque


@dataclass
class BFSResult:
    """Results from BFS traversal with performance metrics"""
    visited_order: List[int]
    distances: Dict[int, int]
    parent: Dict[int, Optional[int]]
    time_taken: float
    vertices_processed: int
    edges_explored: int
    method: str
    num_workers: int


class Graph:
    """Graph representation for BFS"""

    def __init__(self, directed: bool = False):
        """
        Initialize graph.

        Args:
            directed: Whether graph is directed
        """
        self.adj_list = defaultdict(list)
        self.directed = directed
        self.num_vertices = 0

    def add_edge(self, u: int, v: int):
        """Add edge from u to v"""
        self.adj_list[u].append(v)
        if not self.directed:
            self.adj_list[v].append(u)
        self.num_vertices = max(self.num_vertices, u + 1, v + 1)

    def get_neighbors(self, vertex: int) -> List[int]:
        """Get neighbors of a vertex"""
        return self.adj_list.get(vertex, [])

    def num_edges(self) -> int:
        """Get number of edges"""
        count = sum(len(neighbors) for neighbors in self.adj_list.values())
        return count if self.directed else count // 2


class ParallelBFS:
    """
    Parallel Breadth-First Search with level synchronization.
    """

    def __init__(self, graph: Graph, num_workers: int = None):
        """
        Initialize ParallelBFS.

        Args:
            graph: Graph to traverse
            num_workers: Number of worker threads
        """
        self.graph = graph
        self.num_workers = num_workers or mp.cpu_count()
        self.visited = set()
        self.distances = {}
        self.parent = {}
        self.visited_order = []
        self._lock = threading.Lock()
        self.edges_explored = 0

    def _reset(self):
        """Reset traversal state"""
        self.visited = set()
        self.distances = {}
        self.parent = {}
        self.visited_order = []
        self.edges_explored = 0

    def sequential_bfs(self, start: int) -> BFSResult:
        """
        Standard sequential BFS.

        Args:
            start: Starting vertex

        Returns:
            BFSResult with traversal information
        """
        self._reset()
        start_time = time.perf_counter()

        q = deque([start])
        self.visited.add(start)
        self.distances[start] = 0
        self.parent[start] = None
        self.visited_order.append(start)

        while q:
            vertex = q.popleft()

            for neighbor in self.graph.get_neighbors(vertex):
                self.edges_explored += 1
                if neighbor not in self.visited:
                    self.visited.add(neighbor)
                    self.distances[neighbor] = self.distances[vertex] + 1
                    self.parent[neighbor] = vertex
                    self.visited_order.append(neighbor)
                    q.append(neighbor)

        time_taken = time.perf_counter() - start_time

        return BFSResult(
            visited_order=self.visited_order.copy(),
            distances=self.distances.copy(),
            parent=self.parent.copy(),
            time_taken=time_taken,
            vertices_processed=len(self.visited),
            edges_explored=self.edges_explored,
            method="sequential",
            num_workers=1
        )

    def parallel_bfs_level_sync(self, start: int) -> BFSResult:
        """
        Level-synchronous parallel BFS.

        Processes each level in parallel before moving to the next level.
        This maintains BFS ordering and correctness.

        Args:
            start: Starting vertex

        Returns:
            BFSResult with traversal information
        """
        self._reset()
        start_time = time.perf_counter()

        # Initialize with start vertex
        current_level = [start]
        self.visited.add(start)
        self.distances[start] = 0
        self.parent[start] = None
        self.visited_order.append(start)
        level = 0

        while current_level:
            next_level = []
            next_level_lock = threading.Lock()

            def process_vertex(vertex: int):
                """Process a single vertex and discover neighbors"""
                local_neighbors = []

                for neighbor in self.graph.get_neighbors(vertex):
                    with self._lock:
                        self.edges_explored += 1
                        # Check if already visited
                        if neighbor not in self.visited:
                            self.visited.add(neighbor)
                            self.distances[neighbor] = level + 1
                            self.parent[neighbor] = vertex
                            local_neighbors.append(neighbor)

                # Add discovered neighbors to next level
                if local_neighbors:
                    with next_level_lock:
                        next_level.extend(local_neighbors)
                        self.visited_order.extend(local_neighbors)

            # Process current level in parallel
            with concurrent.futures.ThreadPoolExecutor(max_workers=self.num_workers) as executor:
                executor.map(process_vertex, current_level)

            current_level = next_level
            level += 1

        time_taken = time.perf_counter() - start_time

        return BFSResult(
            visited_order=self.visited_order.copy(),
            distances=self.distances.copy(),
            parent=self.parent.copy(),
            time_taken=time_taken,
            vertices_processed=len(self.visited),
            edges_explored=self.edges_explored,
            method="level_sync",
            num_workers=self.num_workers
        )

    def parallel_bfs_concurrent_queue(self, start: int) -> BFSResult:
        """
        Parallel BFS using concurrent queue.

        Workers continuously fetch work from a shared queue.
        Less synchronized than level-sync but may not maintain exact BFS order.

        Args:
            start: Starting vertex

        Returns:
            BFSResult with traversal information
        """
        self._reset()
        start_time = time.perf_counter()

        # Thread-safe queue
        work_queue = queue.Queue()
        work_queue.put(start)
        self.visited.add(start)
        self.distances[start] = 0
        self.parent[start] = None
        self.visited_order.append(start)

        # Track active workers
        active_workers = [0]
        active_lock = threading.Lock()
        done_event = threading.Event()

        def worker():
            """Worker that processes vertices from queue"""
            while not done_event.is_set():
                try:
                    vertex = work_queue.get(timeout=0.001)

                    with active_lock:
                        active_workers[0] += 1

                    # Process vertex
                    for neighbor in self.graph.get_neighbors(vertex):
                        with self._lock:
                            self.edges_explored += 1
                            if neighbor not in self.visited:
                                self.visited.add(neighbor)
                                self.distances[neighbor] = self.distances[vertex] + 1
                                self.parent[neighbor] = vertex
                                self.visited_order.append(neighbor)
                                work_queue.put(neighbor)

                    with active_lock:
                        active_workers[0] -= 1

                    work_queue.task_done()

                except queue.Empty:
                    # Check if all done
                    with active_lock:
                        if work_queue.empty() and active_workers[0] == 0:
                            done_event.set()

        # Start workers
        threads = []
        for _ in range(self.num_workers):
            t = threading.Thread(target=worker)
            t.start()
            threads.append(t)

        # Wait for completion
        for t in threads:
            t.join()

        time_taken = time.perf_counter() - start_time

        return BFSResult(
            visited_order=self.visited_order.copy(),
            distances=self.distances.copy(),
            parent=self.parent.copy(),
            time_taken=time_taken,
            vertices_processed=len(self.visited),
            edges_explored=self.edges_explored,
            method="concurrent_queue",
            num_workers=self.num_workers
        )

    def parallel_bfs_bag_of_tasks(self, start: int) -> BFSResult:
        """
        Parallel BFS using bag-of-tasks pattern.

        Processes vertices in batches for better load balancing.

        Args:
            start: Starting vertex

        Returns:
            BFSResult with traversal information
        """
        self._reset()
        start_time = time.perf_counter()

        current_batch = [start]
        self.visited.add(start)
        self.distances[start] = 0
        self.parent[start] = None
        self.visited_order.append(start)
        distance = 0

        while current_batch:
            # Split batch among workers
            batch_size = max(1, len(current_batch) // self.num_workers)
            batches = [current_batch[i:i + batch_size]
                      for i in range(0, len(current_batch), batch_size)]

            next_batch = []
            next_batch_lock = threading.Lock()

            def process_batch(vertices: List[int]):
                """Process a batch of vertices"""
                local_next = []

                for vertex in vertices:
                    for neighbor in self.graph.get_neighbors(vertex):
                        with self._lock:
                            self.edges_explored += 1
                            if neighbor not in self.visited:
                                self.visited.add(neighbor)
                                self.distances[neighbor] = distance + 1
                                self.parent[neighbor] = vertex
                                local_next.append(neighbor)

                with next_batch_lock:
                    next_batch.extend(local_next)
                    self.visited_order.extend(local_next)

            # Process batches in parallel
            with concurrent.futures.ThreadPoolExecutor(max_workers=self.num_workers) as executor:
                executor.map(process_batch, batches)

            current_batch = next_batch
            distance += 1

        time_taken = time.perf_counter() - start_time

        return BFSResult(
            visited_order=self.visited_order.copy(),
            distances=self.distances.copy(),
            parent=self.parent.copy(),
            time_taken=time_taken,
            vertices_processed=len(self.visited),
            edges_explored=self.edges_explored,
            method="bag_of_tasks",
            num_workers=self.num_workers
        )


def create_random_graph(num_vertices: int, edges_per_vertex: int = 5) -> Graph:
    """Create a random graph for testing"""
    import random
    graph = Graph(directed=False)

    for v in range(num_vertices):
        # Add random edges
        for _ in range(edges_per_vertex):
            neighbor = random.randint(0, num_vertices - 1)
            if neighbor != v:
                graph.add_edge(v, neighbor)

    return graph


def benchmark_all_methods(graph: Graph, start: int, num_workers: int = None) -> dict:
    """Benchmark all BFS methods"""
    results = {}

    methods = ["sequential", "level_sync", "concurrent_queue", "bag_of_tasks"]

    for method in methods:
        bfs = ParallelBFS(graph, num_workers=num_workers)

        if method == "sequential":
            results[method] = bfs.sequential_bfs(start)
        elif method == "level_sync":
            results[method] = bfs.parallel_bfs_level_sync(start)
        elif method == "concurrent_queue":
            results[method] = bfs.parallel_bfs_concurrent_queue(start)
        elif method == "bag_of_tasks":
            results[method] = bfs.parallel_bfs_bag_of_tasks(start)

    return results


def calculate_speedup(seq_time: float, parallel_time: float) -> float:
    """Calculate speedup factor"""
    return seq_time / parallel_time if parallel_time > 0 else 0


# Example usage
if __name__ == "__main__":
    print("=" * 80)
    print("PARALLEL BREADTH-FIRST SEARCH (BFS)")
    print("=" * 80)

    # Test with different graph sizes
    sizes = [(1000, 5), (5000, 10), (10000, 15)]

    for num_vertices, edges_per_vertex in sizes:
        print(f"\n{'='*80}")
        print(f"Testing with {num_vertices:,} vertices, ~{edges_per_vertex} edges per vertex")
        print(f"{'='*80}")

        # Create random graph
        graph = create_random_graph(num_vertices, edges_per_vertex)
        print(f"Graph: {graph.num_vertices:,} vertices, ~{graph.num_edges():,} edges")

        # Benchmark all methods
        results = benchmark_all_methods(graph, start=0)

        print(f"\n{'Method':<20} {'Time (s)':<12} {'Vertices':<12} {'Speedup':<10}")
        print("-" * 80)

        seq_time = results.get("sequential").time_taken if "sequential" in results else 0

        for method, result in results.items():
            speedup = calculate_speedup(seq_time, result.time_taken)

            print(f"{method:<20} {result.time_taken:>10.6f}  "
                  f"{result.vertices_processed:>10,}  "
                  f"{speedup:>8.2f}x")

    print(f"\n{'='*80}")
    print("WHEN TO USE PARALLEL BFS")
    print(f"{'='*80}")
    print("""
Parallel BFS is beneficial when:

✓ Large graphs (>10,000 vertices)
✓ High branching factor (many neighbors per vertex)
✓ Multiple CPU cores available
✓ Graph fits in memory

Key strategies:
1. Level-synchronous: Best for correctness, maintains BFS order
2. Concurrent queue: Good balance of performance and simplicity
3. Bag-of-tasks: Best load balancing for irregular graphs

Performance characteristics:
- Speedup limited by graph diameter (sequential depth)
- Best for "wide" graphs (high branching factor)
- Worse for "deep" graphs (low branching factor)
- Synchronization overhead for small graphs

Challenges:
- Thread contention on visited set
- Load imbalance for irregular graphs
- Memory bandwidth limitations
- Synchronization overhead

Best practices:
1. Use level-synchronous for exact BFS ordering
2. Use concurrent queue for better load balance
3. Optimize visited set (lock-free data structures)
4. Consider graph partitioning for very large graphs
5. Profile to identify bottlenecks

Applications:
- Social network analysis
- Web crawling
- Network routing
- Shortest path finding
- Connected components
    """)

    print("\nDemonstration complete!")
